windows/src/engine/kmtip/keys.cpp

```diff
@@ -260,35 +259,46 @@ HRESULT CKMTipTextService::_PreserveAltKeys(ITfKeystrokeMgr *pKeystrokeMgr)   //
   GetKeyboardPreservedKeys pGetKeyboardPreservedKeys = (GetKeyboardPreservedKeys)GetProcAddress(_hKeyman, "GetKeyboardPreservedKeys");
   if(pGetKeyboardPreservedKeys == NULL) 
   {
+    DebugLastError(L"GetProcAddress");
     return E_FAIL;
   }
 
   if(!(*pGetKeyboardPreservedKeys)(NULL, &_cPreservedKeyCount))
   {
+    Log(L"GetKeyboardPreservedKeys failed");
     return E_FAIL;
   }
 
   _PreservedKeys = new PreservedKey[_cPreservedKeyCount];
 
   if(!(*pGetKeyboardPreservedKeys)(&_PreservedKeys, &_cPreservedKeyCount))
   {
+    Log(L"GetKeyboardPreservedKeys(2nd) failed");
     return E_FAIL;
   }
 
-  /*TF_PRESERVEDKEY pkey;
-    pkey.uModifiers = TF_MOD_ALT;
-    pkey.uVKey = 0x41;*/
-
   HRESULT hr = S_OK;
 
   for(size_t i = 0; i < _cPreservedKeyCount; i++)
   {
     hr = pKeystrokeMgr->PreserveKey(_tfClientId, _PreservedKeys[i].guid, &_PreservedKeys[i].key, NULL, 0);
-    if(hr != S_OK)
-      Log(L"Failed to preserve key %d, %x", i, hr);
+    if (hr != S_OK) {
+      if(_PreservedKeys[i].key.uVKey < 256)
+        Log(L"Failed to preserve key %d: %s %x, %x", i, VKeyNames[_PreservedKeys[i].key.uVKey], _PreservedKeys[i].key.uModifiers, hr);
+      else
+        Log(L"Failed to preserve key %d: %x %x, %x", i, _PreservedKeys[i].key.uVKey, _PreservedKeys[i].key.uModifiers, hr);
+    }
+#ifdef _DEBUG_LOGKEY
+    else {
+      if (_PreservedKeys[i].key.uVKey < 256)
+        Log(L"Preserved key %d: %s %x, %x", i, VKeyNames[_PreservedKeys[i].key.uVKey], _PreservedKeys[i].key.uModifiers, hr);
+      else
+        Log(L"Preserved key %d: %x %x, %x", i, _PreservedKeys[i].key.uVKey, _PreservedKeys[i].key.uModifiers, hr);
+    }
+#endif
   }    
 
-  return hr;
+  return S_OK;
```

Comment:
Why didn't you keep this `return hr;` for when hr != S_OK?

Reply:
Even when some keys fail to preserve, we want to do a best effort. The failing keys are logged.